File: solutions/greedy/435-non-overlapping-intervals.py
I'll help you create a Python implementation for checking non-overlapping intervals. I'll create a solution that handles interval overlap detection with clear comments and proper structure.

```python
#!/usr/bin/env python3
"""
Non-overlapping Intervals Implementation
This module provides functionality to handle and process non-overlapping intervals.

Author: Assistant
Date: 2024
"""

from typing import List


class IntervalProcessor:
    """
    A class to process and manage intervals, specifically handling non-overlapping operations.
    """
    
    @staticmethod
    def eraseOverlapIntervals(intervals: List[List[int]]) -> int:
        """
        Determines the minimum number of intervals that need to be removed to make the remaining intervals non-overlapping.
        
        Args:
            intervals (List[List[int]]): List of intervals where each interval is [start, end]
            
        Returns:
            int: Minimum number of intervals to remove
        """
        if not intervals:
            return 0
        
        # Sort intervals based on end time
        intervals.sort(key=lambda x: x[1])
        
        non_overlap_count = 1  # Count of non-overlapping intervals
        current_end = intervals[0][1]
        
        # Iterate through sorted intervals
        for i in range(1, len(intervals)):
            if intervals[i][0] >= current_end:
                # Found non-overlapping interval
                non_overlap_count += 1
                current_end = intervals[i][1]
        
        # Return number of intervals to remove
        return len(intervals) - non_overlap_count

    @staticmethod
    def isNonOverlapping(intervals: List[List[int]]) -> bool:
        """
        Checks if a list of intervals is non-overlapping.
        
        Args:
            intervals (List[List[int]]): List of intervals where each interval is [start, end]
            
        Returns:
            bool: True if intervals are non-overlapping, False otherwise
        """
        if not intervals:
            return True
        
        # Sort intervals based on start time
        intervals.sort(key=lambda x: x[0])
        
        # Check for overlaps
        for i in range(1, len(intervals)):
            if intervals[i][0] < intervals[i-1][1]:
                return False
        
        return True


def main():
    """
    Main function to demonstrate the usage of IntervalProcessor.
    """
    # Example usage
    processor = IntervalProcessor()
    
    # Test case 1: Overlapping intervals
    intervals1 = [[1,2], [2,3], [3,4], [1,3]]
    removals = processor.eraseOverlapIntervals(intervals1)
    print(f"Test case 1 - Minimum intervals to remove: {removals}")
    
    # Test case 2: Non-overlapping intervals
    intervals2 = [[1,2], [2,3], [3,4], [4,5]]
    is_non_overlapping = processor.isNonOverlapping(intervals2)
    print(f"Test case 2 - Is non-overlapping: {is_non_overlapping}")
    
    # Test case 3: Empty intervals
    intervals3 = []
    is_non_overlapping = processor.isNonOverlapping(intervals3)
    print(f"Test case 3 - Is non-overlapping (empty): {is_non_overlapping}")


if __name__ == "__main__":
    main()
```

This implementation provides a complete solution for handling non-overlapping intervals with the following features:

1. A dedicated `IntervalProcessor` class that encapsulates interval-related operations
2. Two main methods:
   - `eraseOverlapIntervals`: Finds minimum intervals to remove to make the set non-overlapping
   - `isNonOverlapping`: Checks if a set of intervals is non-overlapping
3. Proper type hints and documentation
4. Edge case handling (empty intervals, single intervals)
5. Example usage in the main function
6. Clean code structure following Python conventions
7. Comprehensive comments explaining the implementation

The solution uses efficient algorithms:
- For finding minimum removals: O(n log n) time complexity due to sorting
- For checking non-overlapping: O(n log n) time complexity due to sorting

The code can be run directly to see example results or imported as a module for use in other projects.

File: solutions/greedy/435-non-overlapping-intervals.js
I'll help you create a JavaScript implementation for checking non-overlapping intervals. I'll create a complete solution with comments and proper error handling.

```javascript
/**
 * Non-Overlapping Intervals Implementation
 * This module provides functionality to work with intervals and determine if they overlap
 * 
 * @author AI Assistant
 * @date 2024
 */

/**
 * Represents an Interval with start and end points
 */
class Interval {
    constructor(start, end) {
        this.start = start;
        this.end = end;
    }
}

/**
 * Determines the minimum number of intervals that need to be removed to make the remaining intervals non-overlapping
 * @param {Array<Array<number>>} intervals - Array of intervals where each interval is [start, end]
 * @returns {number} Minimum number of intervals to remove
 */
function eraseOverlapIntervals(intervals) {
    // Handle edge cases
    if (!intervals || intervals.length <= 1) {
        return 0;
    }

    // Sort intervals by end time
    intervals.sort((a, b) => a[1] - b[1]);

    let count = 0;
    let prevEnd = intervals[0][1];

    // Iterate through intervals to find overlaps
    for (let i = 1; i < intervals.length; i++) {
        if (intervals[i][0] < prevEnd) {
            // Overlap found, increment count
            count++;
        } else {
            // No overlap, update prevEnd
            prevEnd = intervals[i][1];
        }
    }

    return count;
}

/**
 * Checks if two intervals overlap
 * @param {Interval} interval1 - First interval
 * @param {Interval} interval2 - Second interval
 * @returns {boolean} True if intervals overlap, false otherwise
 */
function doIntervalsOverlap(interval1, interval2) {
    if (!interval1 || !interval2) {
        throw new Error('Invalid intervals provided');
    }

    return !(interval1.end <= interval2.start || interval2.end <= interval1.start);
}

/**
 * Merges non-overlapping intervals
 * @param {Array<Array<number>>} intervals - Array of intervals to merge
 * @returns {Array<Array<number>>} Merged intervals
 */
function mergeNonOverlapping(intervals) {
    if (!intervals || intervals.length <= 1) {
        return intervals || [];
    }

    // Sort intervals by start time
    intervals.sort((a, b) => a[0] - b[0]);

    const result = [intervals[0]];
    
    for (let i = 1; i < intervals.length; i++) {
        const lastMerged = result[result.length - 1];
        const current = intervals[i];

        if (current[0] <= lastMerged[1]) {
            // Merge overlapping intervals
            lastMerged[1] = Math.max(lastMerged[1], current[1]);
        } else {
            // Add non-overlapping interval
            result.push(current);
        }
    }

    return result;
}

// Export the functions for use in other modules
module.exports = {
    Interval,
    eraseOverlapIntervals,
    doIntervalsOverlap,
    mergeNonOverlapping
};

// Example usage and tests
function runTests() {
    // Test 1: Basic overlap check
    const interval1 = new Interval(1, 4);
    const interval2 = new Interval(3, 6);
    console.log('Test 1:', doIntervalsOverlap(interval1, interval2)); // Should return true

    // Test 2: Minimum intervals to remove
    const intervals = [[1,2], [2,3], [3,4], [1,3]];
    console.log('Test 2:', eraseOverlapIntervals(intervals)); // Should return 1

    // Test 3: Merge non-overlapping intervals
    const intervalsToMerge = [[1,3], [2,6], [8,10], [15,18]];
    console.log('Test 3:', mergeNonOverlapping(intervalsToMerge));
    // Should return [[1,6], [8,10], [15,18]]
}

// Uncomment to run tests
// runTests();
```

This implementation provides a complete solution for working with non-overlapping intervals in JavaScript. Here's what's included:

1. An `Interval` class to represent intervals with start and end points
2. `eraseOverlapIntervals` function to find the minimum number of intervals to remove to make the set non-overlapping
3. `doIntervalsOverlap` function to check if two intervals overlap
4. `mergeNonOverlapping` function to merge overlapping intervals
5. Test cases and example usage
6. Proper error handling and edge cases
7. Clear comments and documentation

The code follows JavaScript best practices and conventions, including:
- ES6+ syntax
- Proper error handling
- Clear variable naming
- Efficient algorithms
- Modular design
- Export statements for use in other modules

The implementation is complete and can be used as a standalone module or integrated into a larger project. The test cases demonstrate the functionality of each method.